File: tools/pvmi/proc_pid_status_metrics.py
#! /usr/bin/env python3

# support for pvmi/proc_pid_metrics.go testing

from typing import Optional

import procfs
from procfs.proc_status import UidGidIndex
from tools_common import sanitize_label_value

from .common import Metric, register_metrics_fn

metrics_fn_map = {}


@register_metrics_fn(
    metrics_fn_map,
    [
        "Name",
        "TGID",
        ("UIDs", UidGidIndex.REAL_ID),
        ("UIDs", UidGidIndex.EFFECTIVE_ID),
        ("UIDs", UidGidIndex.SAVED_ID),
        ("UIDs", UidGidIndex.FILESYSTEM_ID),
        ("GIDs", UidGidIndex.REAL_ID),
        ("GIDs", UidGidIndex.EFFECTIVE_ID),
        ("GIDs", UidGidIndex.SAVED_ID),
        ("GIDs", UidGidIndex.FILESYSTEM_ID),
    ],
)
def proc_pid_status_info_metric(
    procStatus: procfs.ProcStatus,
    common_labels: str = "",
    ts: Optional[int] = None,
    _val: int = 1,
) -> Metric:
    return Metric(
        metric=(
            "proc_pid_status_info{"
            + (
                common_labels
                + ("," if common_labels else "")
                + f'''name="{sanitize_label_value(procStatus.Name)}"'''
                + f''',tgid="{procStatus.TGID}"'''
                + f''',real_uid="{procStatus.UIDs[UidGidIndex.REAL_ID]}"'''
                + f''',effective_uid="{procStatus.UIDs[UidGidIndex.EFFECTIVE_ID]}"'''
                + f''',saved_uid="{procStatus.UIDs[UidGidIndex.SAVED_ID]}"'''
                + f''',filesystem_uid="{procStatus.UIDs[UidGidIndex.FILESYSTEM_ID]}"'''
                + f''',real_gid="{procStatus.GIDs[UidGidIndex.REAL_ID]}"'''
                + f''',effective_gid="{procStatus.GIDs[UidGidIndex.EFFECTIVE_ID]}"'''
                + f''',saved_gid="{procStatus.GIDs[UidGidIndex.SAVED_ID]}"'''
                + f''',filesystem_gid="{procStatus.GIDs[UidGidIndex.FILESYSTEM_ID]}"'''
            )
            + "}"
        ),
        val=_val,
        ts=ts if ts is not None else procStatus._ts,
    )


@register_metrics_fn(metrics_fn_map, "VmPeak")
def proc_pid_status_vm_peak_metric(
    procStatus: procfs.ProcStatus,
    common_labels: str = "",
    ts: Optional[int] = None,
) -> Metric:
    return Metric(
        metric="proc_pid_status_vm_peak" + "{" + common_labels + "}",
        val=procStatus.VmPeak,
        ts=ts if ts is not None else procStatus._ts,
    )


@register_metrics_fn(metrics_fn_map, "VmSize")
def proc_pid_status_vm_peak_metric(
    procStatus: procfs.ProcStatus,
    common_labels: str = "",
    ts: Optional[int] = None,
) -> Metric:
    return Metric(
        metric="proc_pid_status_vm_size" + "{" + common_labels + "}",
        val=procStatus.VmSize,
        ts=ts if ts is not None else procStatus._ts,
    )


@register_metrics_fn(metrics_fn_map, "VmLck")
def proc_pid_status_vm_lck_metric(
    procStatus: procfs.ProcStatus,
    common_labels: str = "",
    ts: Optional[int] = None,
) -> Metric:
    return Metric(
        metric="proc_pid_status_vm_lck" + "{" + common_labels + "}",
        val=procStatus.VmLck,
        ts=ts if ts is not None else procStatus._ts,
    )


@register_metrics_fn(metrics_fn_map, "VmPin")
def proc_pid_status_vm_lck_metric(
    procStatus: procfs.ProcStatus,
    common_labels: str = "",
    ts: Optional[int] = None,
) -> Metric:
    return Metric(
        metric="proc_pid_status_vm_pin" + "{" + common_labels + "}",
        val=procStatus.VmPin,
        ts=ts if ts is not None else procStatus._ts,
    )


@register_metrics_fn(metrics_fn_map, "VmHWM")
def proc_pid_status_vm_hwm_metric(
    procStatus: procfs.ProcStatus,
    common_labels: str = "",
    ts: Optional[int] = None,
) -> Metric:
    return Metric(
        metric="proc_pid_status_vm_hwm" + "{" + common_labels + "}",
        val=procStatus.VmHWM,
        ts=ts if ts is not None else procStatus._ts,
    )


@register_metrics_fn(metrics_fn_map, "VmRSS")
def proc_pid_status_vm_rss_metric(
    procStatus: procfs.ProcStatus,
    common_labels: str = "",
    ts: Optional[int] = None,
) -> Metric:
    return Metric(
        metric="proc_pid_status_vm_rss" + "{" + common_labels + "}",
        val=procStatus.VmRSS,
        ts=ts if ts is not None else procStatus._ts,
    )


@register_metrics_fn(metrics_fn_map, "RssAnon")
def proc_pid_status_vm_rss_anon_metric(
    procStatus: procfs.ProcStatus,
    common_labels: str = "",
    ts: Optional[int] = None,
) -> Metric:
    return Metric(
        metric="proc_pid_status_vm_rss_anon" + "{" + common_labels + "}",
        val=procStatus.RssAnon,
        ts=ts if ts is not None else procStatus._ts,
    )


@register_metrics_fn(metrics_fn_map, "RssFile")
def proc_pid_status_vm_rss_file_metric(
    procStatus: procfs.ProcStatus,
    common_labels: str = "",
    ts: Optional[int] = None,
) -> Metric:
    return Metric(
        metric="proc_pid_status_vm_rss_file" + "{" + common_labels + "}",
        val=procStatus.RssFile,
        ts=ts if ts is not None else procStatus._ts,
    )


@register_metrics_fn(metrics_fn_map, "RssShmem")
def proc_pid_status_vm_rss_shmem_metric(
    procStatus: procfs.ProcStatus,
    common_labels: str = "",
    ts: Optional[int] = None,
) -> Metric:
    return Metric(
        metric="proc_pid_status_vm_rss_shmem" + "{" + common_labels + "}",
        val=procStatus.RssShmem,
        ts=ts if ts is not None else procStatus._ts,
    )


@register_metrics_fn(metrics_fn_map, "VmData")
def proc_pid_status_vm_data_metric(
    procStatus: procfs.ProcStatus,
    common_labels: str = "",
    ts: Optional[int] = None,
) -> Metric:
    return Metric(
        metric="proc_pid_status_vm_data" + "{" + common_labels + "}",
        val=procStatus.VmData,
        ts=ts if ts is not None else procStatus._ts,
    )


@register_metrics_fn(metrics_fn_map, "VmStk")
def proc_pid_status_vm_stk_metric(
    procStatus: procfs.ProcStatus,
    common_labels: str = "",
    ts: Optional[int] = None,
) -> Metric:
    return Metric(
        metric="proc_pid_status_vm_stk" + "{" + common_labels + "}",
        val=procStatus.VmStk,
        ts=ts if ts is not None else procStatus._ts,
    )


@register_metrics_fn(metrics_fn_map, "VmExe")
def proc_pid_status_vm_exe_metric(
    procStatus: procfs.ProcStatus,
    common_labels: str = "",
    ts: Optional[int] = None,
) -> Metric:
    return Metric(
        metric="proc_pid_status_vm_exe" + "{" + common_labels + "}",
        val=procStatus.VmExe,
        ts=ts if ts is not None else procStatus._ts,
    )


@register_metrics_fn(metrics_fn_map, "VmLib")
def proc_pid_status_vm_lib_metric(
    procStatus: procfs.ProcStatus,
    common_labels: str = "",
    ts: Optional[int] = None,
) -> Metric:
    return Metric(
        metric="proc_pid_status_vm_lib" + "{" + common_labels + "}",
        val=procStatus.VmLib,
        ts=ts if ts is not None else procStatus._ts,
    )


@register_metrics_fn(metrics_fn_map, "VmPTE")
def proc_pid_status_vm_pte_metric(
    procStatus: procfs.ProcStatus,
    common_labels: str = "",
    ts: Optional[int] = None,
) -> Metric:
    return Metric(
        metric="proc_pid_status_vm_pte" + "{" + common_labels + "}",
        val=procStatus.VmPTE,
        ts=ts if ts is not None else procStatus._ts,
    )


@register_metrics_fn(metrics_fn_map, "VmPMD")
def proc_pid_status_vm_pmd_metric(
    procStatus: procfs.ProcStatus,
    common_labels: str = "",
    ts: Optional[int] = None,
) -> Metric:
    return Metric(
        metric="proc_pid_status_vm_pmd" + "{" + common_labels + "}",
        val=procStatus.VmPMD,
        ts=ts if ts is not None else procStatus._ts,
    )


@register_metrics_fn(metrics_fn_map, "VmSwap")
def proc_pid_status_vm_swap_metric(
    procStatus: procfs.ProcStatus,
    common_labels: str = "",
    ts: Optional[int] = None,
) -> Metric:
    return Metric(
        metric="proc_pid_status_vm_swap" + "{" + common_labels + "}",
        val=procStatus.VmSwap,
        ts=ts if ts is not None else procStatus._ts,
    )


@register_metrics_fn(metrics_fn_map, "HugetlbPages")
def proc_pid_status_hugetbl_pages_metric(
    procStatus: procfs.ProcStatus,
    common_labels: str = "",
    ts: Optional[int] = None,
) -> Metric:
    return Metric(
        metric="proc_pid_status_hugetbl_pages" + "{" + common_labels + "}",
        val=procStatus.HugetlbPages,
        ts=ts if ts is not None else procStatus._ts,
    )


@register_metrics_fn(metrics_fn_map, "VoluntaryCtxtSwitches")
def proc_pid_status_voluntary_ctxt_switches_metric(
    procStatus: procfs.ProcStatus,
    common_labels: str = "",
    ts: Optional[int] = None,
) -> Metric:
    return Metric(
        metric="proc_pid_status_voluntary_ctxt_switches" + "{" + common_labels + "}",
        val=procStatus.VoluntaryCtxtSwitches,
        ts=ts if ts is not None else procStatus._ts,
    )


@register_metrics_fn(metrics_fn_map, "NonVoluntaryCtxtSwitches")
def proc_pid_status_nonvoluntary_ctxt_switches_metric(
    procStatus: procfs.ProcStatus,
    common_labels: str = "",
    ts: Optional[int] = None,
) -> Metric:
    return Metric(
        metric="proc_pid_status_nonvoluntary_ctxt_switches" + "{" + common_labels + "}",
        val=procStatus.NonVoluntaryCtxtSwitches,
        ts=ts if ts is not None else procStatus._ts,
    )

File: pvmi/proc_pid_status_metrics.go
// proc_pid_status_* metrics utils.

package pvmi

import (
	"bytes"
	"fmt"

	"github.com/prometheus/procfs"
)

const (
	PROC_PID_STATUS_INFO_METRIC_NAME               = "proc_pid_status_info"
	PROC_PID_STATUS_INFO_NAME_LABEL_NAME           = "name"
	PROC_PID_STATUS_INFO_TGID_LABEL_NAME           = "tgid"
	PROC_PID_STATUS_INFO_REAL_UID_LABEL_NAME       = "real_uid"
	PROC_PID_STATUS_INFO_EFFECTIVE_UID_LABEL_NAME  = "effective_uid"
	PROC_PID_STATUS_INFO_SAVED_UID_LABEL_NAME      = "saved_uid"
	PROC_PID_STATUS_INFO_FILESYSTEM_UID_LABEL_NAME = "filesystem_uid"
	PROC_PID_STATUS_INFO_REAL_GID_LABEL_NAME       = "real_gid"
	PROC_PID_STATUS_INFO_EFFECTIVE_GID_LABEL_NAME  = "effective_gid"
	PROC_PID_STATUS_INFO_SAVED_GID_LABEL_NAME      = "saved_gid"
	PROC_PID_STATUS_INFO_FILESYSTEM_GID_LABEL_NAME = "filesystem_gid"

	PROC_PID_STATUS_VM_PEAK_METRIC_NAME = "proc_pid_status_vm_peak"

	PROC_PID_STATUS_VM_SIZE_METRIC_NAME = "proc_pid_status_vm_size"

	PROC_PID_STATUS_VM_LCK_METRIC_NAME = "proc_pid_status_vm_lck"

	PROC_PID_STATUS_VM_PIN_METRIC_NAME = "proc_pid_status_vm_pin"

	PROC_PID_STATUS_VM_HWM_METRIC_NAME = "proc_pid_status_vm_hwm"

	PROC_PID_STATUS_VM_RSS_METRIC_NAME = "proc_pid_status_vm_rss"

	PROC_PID_STATUS_VM_RSS_ANON_METRIC_NAME = "proc_pid_status_vm_rss_anon"

	PROC_PID_STATUS_VM_RSS_FILE_METRIC_NAME = "proc_pid_status_vm_rss_file"

	PROC_PID_STATUS_VM_RSS_SHMEM_METRIC_NAME = "proc_pid_status_vm_rss_shmem"

	PROC_PID_STATUS_VM_DATA_METRIC_NAME = "proc_pid_status_vm_data"

	PROC_PID_STATUS_VM_STK_METRIC_NAME = "proc_pid_status_vm_stk"

	PROC_PID_STATUS_VM_EXE_METRIC_NAME = "proc_pid_status_vm_exe"

	PROC_PID_STATUS_VM_LIB_METRIC_NAME = "proc_pid_status_vm_lib"

	PROC_PID_STATUS_VM_PTE_METRIC_NAME = "proc_pid_status_vm_pte"

	PROC_PID_STATUS_VM_PMD_METRIC_NAME = "proc_pid_status_vm_pmd"

	PROC_PID_STATUS_VM_SWAP_METRIC_NAME = "proc_pid_status_vm_swap"

	PROC_PID_STATUS_HUGETBL_PAGES_METRIC_NAME = "proc_pid_status_hugetbl_pages"

	PROC_PID_STATUS_VOLUNTARY_CTXT_SWITCHES_METRIC_NAME = "proc_pid_status_voluntary_ctxt_switches"

	PROC_PID_STATUS_NONVOLUNTARY_CTXT_SWITCHES_METRIC_NAME = "proc_pid_status_nonvoluntary_ctxt_switches"

	// procfs.ProcStatus.UIDs and .GIDs are [4]string for real, effective, saved
	// and filesystem accordingly:
	PROCFS_STATUS_UID_GID_REAL_INDEX       = 0
	PROCFS_STATUS_UID_GID_EFFECTIVE_INDEX  = 1
	PROCFS_STATUS_UID_GID_SAVED_INDEX      = 2
	PROCFS_STATUS_UID_GID_FILESYSTEM_INDEX = 3
)

var pidStatusInfoMetricFmt = fmt.Sprintf(
	`%s{%%s,%s="%%s",%s="%%d",%s="%%s",%s="%%s",%s="%%s",%s="%%s",%s="%%s",%s="%%s",%s="%%s",%s="%%s"}`,
	PROC_PID_STATUS_INFO_METRIC_NAME,
	PROC_PID_STATUS_INFO_NAME_LABEL_NAME,
	PROC_PID_STATUS_INFO_TGID_LABEL_NAME,
	PROC_PID_STATUS_INFO_REAL_UID_LABEL_NAME,
	PROC_PID_STATUS_INFO_EFFECTIVE_UID_LABEL_NAME,
	PROC_PID_STATUS_INFO_SAVED_UID_LABEL_NAME,
	PROC_PID_STATUS_INFO_FILESYSTEM_UID_LABEL_NAME,
	PROC_PID_STATUS_INFO_REAL_GID_LABEL_NAME,
	PROC_PID_STATUS_INFO_EFFECTIVE_GID_LABEL_NAME,
	PROC_PID_STATUS_INFO_SAVED_GID_LABEL_NAME,
	PROC_PID_STATUS_INFO_FILESYSTEM_GID_LABEL_NAME,
)

// Invoke for new/change in procStatus info:
func updateProcPidStatusInfoMetric(
	pmce *PidMetricsCacheEntry,
	procStatus *procfs.ProcStatus,
	promTs string,
	buf *bytes.Buffer,
	generatedCount *uint64,
) error {
	// Clear previous metric, if any:
	if pmce.ProcPidStatusInfoMetric != "" {
		buf.WriteString(pmce.ProcPidStatusInfoMetric + " 0 " + promTs + "\n")
		if generatedCount != nil {
			*generatedCount += 1
		}
	}
	// Update the metric:
	pmce.ProcPidStatusInfoMetric = fmt.Sprintf(
		pidStatusInfoMetricFmt,
		pmce.CommonLabels,
		procStatus.Name,
		procStatus.TGID,
		procStatus.UIDs[PROCFS_STATUS_UID_GID_REAL_INDEX],
		procStatus.UIDs[PROCFS_STATUS_UID_GID_EFFECTIVE_INDEX],
		procStatus.UIDs[PROCFS_STATUS_UID_GID_SAVED_INDEX],
		procStatus.UIDs[PROCFS_STATUS_UID_GID_FILESYSTEM_INDEX],
		procStatus.GIDs[PROCFS_STATUS_UID_GID_REAL_INDEX],
		procStatus.GIDs[PROCFS_STATUS_UID_GID_EFFECTIVE_INDEX],
		procStatus.GIDs[PROCFS_STATUS_UID_GID_SAVED_INDEX],
		procStatus.GIDs[PROCFS_STATUS_UID_GID_FILESYSTEM_INDEX],
	)
	return nil
}
